File: prog_<1>.py
class stack :
    def __init__(self):
        self.stack = []
        self.size = int(len(self.stack))
        self.top = -1
    def push(self,element):
        self.stack.append(element)
        self.top += 1

    def pop(self):
        if self.top == -1:
            return None

        removed_element = self.stack[self.top]
        self.stack = self.stack[:self.top]
        self.top -= 1
        return removed_element
    def peek(self):
        return self.stack[self.top]

    def is_empty(self):
        return self.top == -1

def evaluate(expression):
    string = stack()
    endmap = {")": "(", "]": "[", "}": "{"}
    for c in expression:
        if c == "(" or c == "[" or c == "{" :
            string.push(c)
        elif c == ")" or c == "]" or c == "}" :
            if string.is_empty() :
                return False
            if string.pop() != endmap[c]:
                return False

    return string.is_empty()

if __name__ == "__main__":
    expression1 = "a+(b-c)*(d"
    print("correct" if evaluate(expression1) else "false" )#the answer I got was False
    expression2 = "m + [a − b ∗ (c + d ∗ {m)]"
    print("correct" if evaluate(expression2) else "false")#the answer I got was False
    expression3 ="a+(b-c)"
    print("correct" if evaluate(expression3) else "false")#the answer I got was correct

File: Questions/prog_<1>.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

#define MAX 1000

struct stack {
    char items[MAX];
    int top;
};

void push(struct stack *s, char element) {
    s->items[++(s->top)] = element;
}

char pop(struct stack *s) {
    if (s->top == -1) {
        return '\0';
    }
    return s->items[(s->top)--];
}

bool is_empty(struct stack *s) {
    return s->top == -1;
}

bool evaluate(char* expression) {
    struct stack string;
    string.top = -1;

    for (int i = 0; expression[i] != '\0'; i++) {
        char c = expression[i];

        if (c == '(' || c == '[' || c == '{') {
            push(&string, c);
        } 
        else if (c == ')' || c == ']' || c == '}') {
            if (is_empty(&string)) {
                return false;
            }

            char top_element = pop(&string);
            
            // Implementation of your endmap logic
            if (c == ')' && top_element != '(') return false;
            if (c == ']' && top_element != '[') return false;
            if (c == '}' && top_element != '{') return false;
        }
    }

    return is_empty(&string);
}

int main() {
    char expression1[] = "a+(b-c)*(d";
    printf("%s\n", evaluate(expression1) ? "correct" : "false");

    char expression2[] = "m + [a - b * (c + d * {m)]";
    printf("%s\n", evaluate(expression2) ? "correct" : "false");

    char expression3[] = "a+(b-c)";
    printf("%s\n", evaluate(expression3) ? "correct" : "false");

    return 0;
}
